pkg/gce-pd-csi-driver/controller.go

```diff
@@ -331,11 +338,154 @@ func (gceCS *GCEControllerServer) CreateVolume(ctx context.Context, req *csi.Cre
+	}
+	prefZones, err := getZonesFromTopology(top.GetPreferred())
+	if err != nil {
+		return nil, fmt.Errorf("could not get zones from preferred topology: %w", err)
+	}
+	reqZones, err := getZonesFromTopology(top.GetRequisite())
+	if err != nil {
+		return nil, fmt.Errorf("could not get zones from requisite topology: %w", err)
+	}
+	prefSet := sets.NewString(prefZones...)
+	reqSet := sets.NewString(reqZones...)
+	prefAndReqSet := prefSet.Union(reqSet)
+	availableZones := prefAndReqSet.List()
+	if prefAndReqSet.Len() == 0 {
+		// If there are no specified zones, this means that there were no aggregate
+		// zones (eg: no nodes running) in the cluster
+		availableZones = gceCS.fallbackRequisiteZones
+	}
+
+	supportedZones, err := gceCS.getSupportedZonesForPDType(ctx, availableZones, params.DiskType)
+	if err != nil {
+		return nil, fmt.Errorf("could not get supported zones for disk type %v from zone list %v: %w", params.DiskType, prefAndReqSet.List(), err)
+	}
+
+	return supportedZones, nil
+}
+
+func (gceCS *GCEControllerServer) createMultiZoneDisk(ctx context.Context, req *csi.CreateVolumeRequest, params common.DiskParameters) (*csi.CreateVolumeResponse, error) {
+	// Determine the zones that are needed.
+	var err error
+
+	// For multi-zone, we either select:
+	// 1) The zones specified in requisite topology requirements
+	// 2) All zones in the region that are compatible with the selected disk type
+	zones, err := gceCS.getMultiZoneProvisioningZones(ctx, req, params)
+	if err != nil {
+		return nil, err
+	}
+
+	multiZoneVolKey := meta.ZonalKey(req.GetName(), common.MultiZoneValue)
+	volumeID, err := common.KeyToVolumeID(multiZoneVolKey, gceCS.CloudProvider.GetDefaultProject())
+	if err != nil {
+		return nil, err
+	}
+	if acquired := gceCS.volumeLocks.TryAcquire(volumeID); !acquired {
+		return nil, status.Errorf(codes.Aborted, common.VolumeOperationAlreadyExistsFmt, volumeID)
+	}
+	defer gceCS.volumeLocks.Release(volumeID)
+
+	createDiskErrs := []error{}
+	createdDisks := make([]*gce.CloudDisk, 0, len(zones))
+	for _, zone := range zones {
+		volKey := meta.ZonalKey(req.GetName(), zone)
+		klog.V(4).Infof("Creating single zone disk for zone %q and volume: %v", zone, volKey)
+		disk, err := gceCS.createSingleDisk(ctx, req, params, volKey, []string{zone})
+		if err != nil {
+			createDiskErrs = append(createDiskErrs, err)
+			continue
+		}
+
+		readonly, _ := getReadOnlyFromCapabilities(req.GetVolumeCapabilities())
+		updateAccessModeErr := gceCS.updateAccessModeIfNecessary(ctx, volKey, disk, readonly)
+		if updateAccessModeErr != nil {
+			createDiskErrs = append(createDiskErrs, updateAccessModeErr)
+			continue
+		}
+
+		createdDisks = append(createdDisks, disk)
+	}
+	if len(createDiskErrs) == 1 {
+		return nil, common.LoggedError("Failed to create disk: ", createDiskErrs[0])
+	} else if len(createDiskErrs) > 1 {
+		// Return a non-final error here, since one of the errors could be a non-final.
```

Comment:
It might be a common case that len(createDiskErrs) == len(zones) and all the errors are the same, and final. eg some bad project configuration. Endlessly retrying that might be a problem?

Reply:
Hmm, good call. Let me add some error merging logic here.

Reply:
Updated to check if the error is one of the retryable codes.